// @mui/material-next/node/Slider/Slider.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SliderValueLabel = exports.SliderTrack = exports.SliderThumb = exports.SliderRoot = exports.SliderRail = exports.SliderMarkLabel = exports.SliderMark = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _clsx = _interopRequireDefault(require("clsx"));
var _utils = require("@mui/utils");
var _base = require("@mui/base");
var _useSlider = require("@mui/base/useSlider");
var _system = require("@mui/system");
var _useThemeProps = _interopRequireDefault(require("../styles/useThemeProps"));
var _styled = _interopRequireDefault(require("../styles/styled"));
var _useTheme = _interopRequireDefault(require("../styles/useTheme"));
var _shouldSpreadAdditionalProps = _interopRequireDefault(require("../utils/shouldSpreadAdditionalProps"));
var _sliderClasses = _interopRequireWildcard(require("./sliderClasses"));
var _useSliderElementsOverlap = _interopRequireDefault(require("./useSliderElementsOverlap"));
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["aria-label", "aria-valuetext", "aria-labelledby", "component", "color", "classes", "className", "disableSwap", "disabled", "getAriaLabel", "getAriaValueText", "marks", "max", "min", "name", "onChange", "onChangeCommitted", "orientation", "size", "step", "scale", "slotProps", "slots", "tabIndex", "track", "value", "valueLabelDisplay", "valueLabelFormat"];
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function Identity(x) {
  return x;
}
const SliderRoot = exports.SliderRoot = (0, _styled.default)('span', {
  name: 'MuiSlider',
  slot: 'Root',
  overridesResolver: (props, styles) => {
    const {
      ownerState
    } = props;
    return [styles.root, styles[`color${(0, _utils.unstable_capitalize)(ownerState.color || 'primary')}`], ownerState.size !== 'medium' && styles[`size${(0, _utils.unstable_capitalize)(ownerState.size)}`], ownerState.marked && styles.marked, ownerState.orientation === 'vertical' && styles.vertical, ownerState.track === 'inverted' && styles.trackInverted, ownerState.track === false && styles.trackFalse];
  }
})(({
  theme,
  ownerState
}) => {
  const {
    vars: tokens
  } = theme;
  return (0, _extends2.default)({
    '--md-comp-slider-thumb-size': ownerState.size === 'small' ? '12px' : '20px',
    borderRadius: tokens.sys.shape.corner.full,
    boxSizing: 'content-box',
    display: 'inline-block',
    position: 'relative',
    cursor: 'pointer',
    touchAction: 'none',
    color: tokens.sys.color[ownerState.color || 'primary'],
    WebkitTapHighlightColor: 'transparent'
  }, ownerState.orientation === 'horizontal' && (0, _extends2.default)({
    height: 4,
    width: '100%',
    // 40px touch target
    padding: '18px 0'
  }, ownerState.size === 'small' && {
    height: 2
  }, ownerState.marked && {
    marginBottom: 20
  }), ownerState.orientation === 'vertical' && (0, _extends2.default)({
    height: '100%',
    width: 4,
    padding: '0 18px'
  }, ownerState.size === 'small' && {
    width: 2
  }, ownerState.marked && {
    marginRight: 44
  }), {
    '@media print': {
      colorAdjust: 'exact'
    },
    [`&.${_sliderClasses.default.disabled}`]: {
      pointerEvents: 'none',
      cursor: 'default',
      color: tokens.sys.color.outline
    },
    [`&.${_sliderClasses.default.dragging}`]: {
      [`& .${_sliderClasses.default.track}`]: {
        transition: 'none'
      },
      [`& .${_sliderClasses.default.thumb}, .${_sliderClasses.default.valueLabel}`]: {
        transition: theme.transitions.create(['border'], {
          duration: theme.transitions.duration.shortest
        })
      }
    }
  });
});
const SliderRail = exports.SliderRail = (0, _styled.default)('span', {
  name: 'MuiSlider',
  slot: 'Rail',
  overridesResolver: (props, styles) => styles.rail
})(({
  theme: {
    vars: tokens
  },
  ownerState
}) => (0, _extends2.default)({
  display: 'block',
  position: 'absolute',
  borderRadius: 'inherit',
  backgroundColor: tokens.sys.color.surfaceContainerHighest,
  boxShadow: tokens.sys.elevation[0]
}, ownerState.orientation === 'horizontal' && {
  width: '100%',
  height: 'inherit',
  top: '50%',
  transform: 'translateY(-50%)'
}, ownerState.orientation === 'vertical' && {
  height: '100%',
  width: 'inherit',
  left: '50%',
  transform: 'translateX(-50%)'
}, ownerState.track === 'inverted' && {
  backgroundColor: ownerState.disabled ? tokens.sys.color.outline : 'currentColor'
}));
const SliderTrack = exports.SliderTrack = (0, _styled.default)('span', {
  name: 'MuiSlider',
  slot: 'Track',
  overridesResolver: (props, styles) => styles.track
})(({
  theme,
  ownerState
}) => {
  const {
    vars: tokens
  } = theme;
  return (0, _extends2.default)({
    display: 'block',
    position: 'absolute',
    borderRadius: 'inherit',
    backgroundColor: 'currentColor',
    boxShadow: tokens.sys.elevation[0],
    transition: theme.transitions.create(['left', 'width', 'bottom', 'height'], {
      duration: theme.transitions.duration.shortest
    })
  }, ownerState.orientation === 'horizontal' && {
    height: 'inherit',
    top: '50%',
    transform: 'translateY(-50%)'
  }, ownerState.orientation === 'vertical' && {
    width: 'inherit',
    left: '50%',
    transform: 'translateX(-50%)'
  }, ownerState.track === false && {
    display: 'none'
  }, ownerState.track === 'inverted' && {
    backgroundColor: tokens.sys.color.surfaceContainerHighest
  });
});
const SliderThumb = exports.SliderThumb = (0, _styled.default)('span', {
  name: 'MuiSlider',
  slot: 'Thumb',
  shouldForwardProp: prop => (0, _system.shouldForwardProp)(prop) && prop !== 'isOverlapping',
  overridesResolver: (props, styles) => {
    const {
      ownerState
    } = props;
    return [styles.thumb, props.isOverlapping && styles.thumbOverlap, styles[`thumbColor${(0, _utils.unstable_capitalize)(ownerState.color || 'primary')}`], ownerState.size !== 'medium' && styles[`thumbSize${(0, _utils.unstable_capitalize)(ownerState.size)}`]];
  }
})(({
  theme,
  ownerState
}) => {
  const {
    vars: tokens
  } = theme;
  function getBoxShadow(state) {
    return `0px 0px 0px 10px rgba(${tokens.sys.color.primaryChannel} / ${tokens.sys.state[state].stateLayerOpacity})`;
  }
  return (0, _extends2.default)({
    zIndex: 1,
    position: 'absolute',
    height: 'var(--md-comp-slider-thumb-size)',
    width: 'var(--md-comp-slider-thumb-size)',
    boxSizing: 'border-box',
    borderRadius: tokens.sys.shape.corner.full,
    outline: 0,
    backgroundColor: 'currentColor',
    border: '1px solid currentColor',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    transition: theme.transitions.create(['box-shadow', 'left', 'right', 'bottom', 'border'], {
      duration: theme.transitions.duration.shortest
    })
  }, ownerState.orientation === 'horizontal' && {
    top: '50%',
    transform: 'translate(-50%, -50%)'
  }, ownerState.orientation === 'vertical' && {
    left: '50%',
    transform: 'translate(-50%, 50%)'
  }, {
    '&::before': {
      position: 'absolute',
      content: '""',
      borderRadius: 'inherit',
      width: 'var(--md-comp-slider-thumb-size)',
      height: 'var(--md-comp-slider-thumb-size)',
      boxShadow: tokens.sys.elevation[1]
    },
    '&::after': {
      position: 'absolute',
      content: '""',
      borderRadius: '50%',
      // 40px is the hit target
      width: 40,
      height: 40,
      top: '50%',
      left: '50%',
      transform: 'translate(-50%, -50%)'
    },
    '&:hover': {
      boxShadow: getBoxShadow('hover'),
      '@media (hover: none)': {
        boxShadow: 'none'
      }
    },
    [`&.${_sliderClasses.default.focusVisible}`]: {
      boxShadow: getBoxShadow('focus')
    },
    [`&.${_sliderClasses.default.active}`]: {
      zIndex: 2,
      boxShadow: getBoxShadow('pressed')
    },
    [`&.${_sliderClasses.default.thumbOverlap}`]: {
      border: `1px solid ${tokens.ref.palette.common.white}`
    },
    [`&.${_sliderClasses.default.disabled}`]: {
      boxShadow: 'none',
      '&:before': {
        boxShadow: tokens.sys.elevation[0]
      }
    }
  });
});
const SliderValueLabel = exports.SliderValueLabel = (0, _styled.default)('span', {
  name: 'MuiSlider',
  slot: 'ValueLabel',
  shouldForwardProp: prop => (0, _system.shouldForwardProp)(prop) && prop !== 'isOverlapping' && prop !== 'valueLabelDisplay' && prop !== 'valueLabelFormat',
  overridesResolver: (props, styles) => [styles.valueLabel, props.isOverlapping && styles.valueLabelOverlap]
})(({
  theme,
  ownerState
}) => {
  const {
    vars: tokens
  } = theme;
  const letterSpacing = `${theme.sys.typescale.label.medium.tracking / theme.sys.typescale.label.medium.size}rem`;
  const labelStyle = {
    color: ownerState.disabled ? tokens.sys.color.surface : tokens.sys.color[`on${(0, _utils.unstable_capitalize)(ownerState.color || 'primary')}`],
    // paddingLeft compensates letter spacing being added only on the right side
    paddingLeft: letterSpacing
  };
  return (0, _extends2.default)({
    zIndex: 1,
    whiteSpace: 'nowrap',
    fontFamily: tokens.sys.typescale.label.medium.family,
    lineHeight: `calc(${tokens.sys.typescale.label.large.lineHeight} / ${tokens.sys.typescale.label.medium.size})`,
    fontWeight: tokens.sys.typescale.label.medium.weight,
    fontSize: theme.typography.pxToRem(theme.sys.typescale.label.medium.size),
    letterSpacing,
    transition: theme.transitions.create(['transform', 'border'], {
      duration: theme.transitions.duration.shortest
    }),
    position: 'absolute',
    backgroundColor: 'currentColor',
    boxShadow: tokens.sys.elevation[0],
    borderRadius: '50% 50% 50% 0',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    boxSizing: 'border-box',
    border: '1px solid currentColor',
    width: 28,
    height: 28,
    [`&.${_sliderClasses.default.valueLabelOverlap}`]: {
      border: `1px solid ${tokens.ref.palette.common.white}`
    }
  }, ownerState.orientation === 'horizontal' && {
    top: ownerState.size === 'small' ? -32 : -36,
    transform: 'translateY(50%) rotate(-45deg) scale(0)',
    [`& .${_sliderClasses.default.valueLabelLabel}`]: (0, _extends2.default)({
      transform: 'rotate(45deg)'
    }, labelStyle),
    [`&.${_sliderClasses.default.valueLabelOpen}`]: {
      transform: 'translateY(0) rotate(-45deg) scale(1)'
    }
  }, ownerState.orientation === 'vertical' && {
    left: ownerState.size === 'small' ? -32 : -36,
    transform: 'translateX(50%) rotate(225deg) scale(0)',
    [`& .${_sliderClasses.default.valueLabelLabel}`]: (0, _extends2.default)({
      transform: 'rotate(-225deg)'
    }, labelStyle),
    [`&.${_sliderClasses.default.valueLabelOpen}`]: {
      transform: 'translateX(0) rotate(225deg) scale(1)'
    }
  }, ownerState.size === 'small' && {
    fontSize: theme.typography.pxToRem(theme.sys.typescale.label.small.size),
    width: 24,
    height: 24
  });
});
const SliderMark = exports.SliderMark = (0, _styled.default)('span', {
  name: 'MuiSlider',
  slot: 'Mark',
  shouldForwardProp: prop => (0, _system.shouldForwardProp)(prop) && prop !== 'markActive',
  overridesResolver: (props, styles) => {
    const {
      markActive
    } = props;
    return [styles.mark, markActive && styles.markActive];
  }
})(({
  theme: {
    vars: tokens
  },
  ownerState,
  markActive
}) => (0, _extends2.default)({
  position: 'absolute',
  width: 2,
  height: 2,
  borderRadius: tokens.sys.shape.corner.full,
  backgroundColor: tokens.sys.color.onSurfaceVariant,
  opacity: 0.38
}, ownerState.orientation === 'horizontal' && {
  top: '50%',
  transform: 'translate(-1px, -50%)'
}, ownerState.orientation === 'vertical' && {
  left: '50%',
  transform: 'translate(-50%, 1px)'
}, markActive && {
  backgroundColor: tokens.sys.color[`on${(0, _utils.unstable_capitalize)(ownerState.color || 'primary')}`]
}));
const SliderMarkLabel = exports.SliderMarkLabel = (0, _styled.default)('span', {
  name: 'MuiSlider',
  slot: 'MarkLabel',
  shouldForwardProp: prop => (0, _system.shouldForwardProp)(prop) && prop !== 'markLabelActive',
  overridesResolver: (props, styles) => styles.markLabel
})(({
  theme,
  ownerState,
  markLabelActive
}) => {
  const {
    vars: tokens
  } = theme;
  return (0, _extends2.default)({
    fontFamily: tokens.sys.typescale.label.medium.family,
    lineHeight: `calc(${tokens.sys.typescale.label.medium.lineHeight} / ${tokens.sys.typescale.label.medium.size})`,
    fontWeight: tokens.sys.typescale.label.medium.weight,
    fontSize: theme.typography.pxToRem(theme.sys.typescale.label.medium.size),
    letterSpacing: tokens.sys.typescale.label.medium.tracking,
    color: tokens.sys.color.onSurfaceVariant,
    position: 'absolute',
    whiteSpace: 'nowrap'
  }, ownerState.orientation === 'horizontal' && {
    top: 36,
    transform: 'translateX(-50%)',
    '@media (pointer: coarse)': {
      top: 44
    }
  }, ownerState.orientation === 'vertical' && {
    left: 36,
    transform: 'translateY(50%)',
    '@media (pointer: coarse)': {
      left: 44
    }
  }, markLabelActive && {
    color: tokens.sys.color.onSurface
  });
});
const useUtilityClasses = ownerState => {
  const {
    disabled,
    dragging,
    marked,
    orientation,
    track,
    classes,
    color,
    size
  } = ownerState;
  const slots = {
    root: ['root', disabled && 'disabled', dragging && 'dragging', marked && 'marked', orientation === 'vertical' && 'vertical', track === 'inverted' && 'trackInverted', track === false && 'trackFalse', color && `color${(0, _utils.unstable_capitalize)(color)}`, size && `size${(0, _utils.unstable_capitalize)(size)}`],
    rail: ['rail'],
    track: ['track'],
    mark: ['mark'],
    markActive: ['markActive'],
    markLabel: ['markLabel'],
    markLabelActive: ['markLabelActive'],
    valueLabel: ['valueLabel'],
    valueLabelOpen: ['valueLabelOpen'],
    valueLabelLabel: ['valueLabelLabel'],
    valueLabelOverlap: ['valueLabelOverlap'],
    thumb: ['thumb', disabled && 'disabled', size && `thumbSize${(0, _utils.unstable_capitalize)(size)}`, color && `thumbColor${(0, _utils.unstable_capitalize)(color)}`],
    thumbOverlap: ['thumbOverlap'],
    active: ['active'],
    disabled: ['disabled'],
    focusVisible: ['focusVisible']
  };
  return (0, _base.unstable_composeClasses)(slots, _sliderClasses.getSliderUtilityClass, classes);
};
const Slider = /*#__PURE__*/React.forwardRef(function Slider(inProps, ref) {
  var _slots$root, _slots$rail, _slots$track, _slots$thumb, _slots$valueLabel, _slots$mark, _slots$markLabel, _slots$input;
  const props = (0, _useThemeProps.default)({
    props: inProps,
    name: 'MuiSlider'
  });
  const theme = (0, _useTheme.default)();
  const isRtl = theme.direction === 'rtl';
  const {
      'aria-label': ariaLabel,
      'aria-valuetext': ariaValuetext,
      'aria-labelledby': ariaLabelledby,
      component = 'span',
      color = 'primary',
      classes: classesProp,
      className,
      disableSwap = false,
      disabled = false,
      getAriaLabel,
      getAriaValueText,
      marks: marksProp = false,
      max = 100,
      min = 0,
      orientation = 'horizontal',
      size = 'medium',
      step = 1,
      scale = Identity,
      slotProps = {},
      slots = {},
      track = 'normal',
      valueLabelDisplay = 'off',
      valueLabelFormat = Identity
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const propsWithDefaultValues = (0, _extends2.default)({}, props, {
    isRtl,
    max,
    min,
    classes: classesProp,
    disabled,
    disableSwap,
    orientation,
    marks: marksProp,
    color,
    size,
    step,
    scale,
    track,
    valueLabelDisplay,
    valueLabelFormat
  });
  const {
    axisProps,
    getRootProps,
    getHiddenInputProps,
    getThumbProps,
    open,
    active,
    axis,
    focusedThumbIndex,
    range,
    dragging,
    marks,
    values,
    trackOffset,
    trackLeap
  } = (0, _useSlider.useSlider)((0, _extends2.default)({}, propsWithDefaultValues, {
    rootRef: ref
  }));
  const ownerState = (0, _extends2.default)({}, propsWithDefaultValues, {
    marked: marks.length > 0 && marks.some(mark => mark.label),
    dragging,
    focusedThumbIndex
  });
  const classes = useUtilityClasses(ownerState);
  const overlapApi = (0, _useSliderElementsOverlap.default)(axis);
  const lastActiveThumbIndexRef = React.useRef(-1);
  if (focusedThumbIndex !== -1) {
    lastActiveThumbIndexRef.current = focusedThumbIndex;
  } else if (active !== -1) {
    lastActiveThumbIndexRef.current = active;
  }
  const RootSlot = (_slots$root = slots.root) != null ? _slots$root : SliderRoot;
  const RailSlot = (_slots$rail = slots.rail) != null ? _slots$rail : SliderRail;
  const TrackSlot = (_slots$track = slots.track) != null ? _slots$track : SliderTrack;
  const ThumbSlot = (_slots$thumb = slots.thumb) != null ? _slots$thumb : SliderThumb;
  const ValueLabelSlot = (_slots$valueLabel = slots.valueLabel) != null ? _slots$valueLabel : SliderValueLabel;
  const MarkSlot = (_slots$mark = slots.mark) != null ? _slots$mark : SliderMark;
  const MarkLabelSlot = (_slots$markLabel = slots.markLabel) != null ? _slots$markLabel : SliderMarkLabel;
  const InputSlot = (_slots$input = slots.input) != null ? _slots$input : 'input';
  const rootProps = (0, _base.useSlotProps)({
    elementType: RootSlot,
    getSlotProps: getRootProps,
    externalSlotProps: slotProps.root,
    externalForwardedProps: other,
    additionalProps: (0, _extends2.default)({}, (0, _shouldSpreadAdditionalProps.default)(RootSlot) && {
      as: component
    }),
    ownerState,
    className: [classes.root, className]
  });
  const railProps = (0, _base.useSlotProps)({
    elementType: RailSlot,
    externalSlotProps: slotProps.rail,
    ownerState,
    className: classes.rail
  });
  const trackProps = (0, _base.useSlotProps)({
    elementType: TrackSlot,
    externalSlotProps: slotProps.track,
    additionalProps: {
      style: (0, _extends2.default)({}, axisProps[axis].offset(trackOffset), axisProps[axis].leap(trackLeap))
    },
    ownerState,
    className: classes.track
  });
  const thumbProps = (0, _base.useSlotProps)({
    elementType: ThumbSlot,
    getSlotProps: () => getThumbProps({
      onTransitionEnd: overlapApi.onThumbMoved,
      onPointerMove: overlapApi.onThumbMoved
    }),
    externalSlotProps: slotProps.thumb,
    ownerState,
    className: classes.thumb
  });
  const valueLabelProps = (0, _base.useSlotProps)({
    elementType: ValueLabelSlot,
    externalSlotProps: slotProps.valueLabel,
    ownerState,
    className: classes.valueLabel
  });
  const markProps = (0, _base.useSlotProps)({
    elementType: MarkSlot,
    externalSlotProps: slotProps.mark,
    ownerState,
    className: classes.mark
  });
  const markLabelProps = (0, _base.useSlotProps)({
    elementType: MarkLabelSlot,
    externalSlotProps: slotProps.markLabel,
    ownerState,
    className: classes.markLabel
  });
  const inputSliderProps = (0, _base.useSlotProps)({
    elementType: InputSlot,
    getSlotProps: getHiddenInputProps,
    externalSlotProps: slotProps.input,
    ownerState
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(RootSlot, (0, _extends2.default)({}, rootProps, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(RailSlot, (0, _extends2.default)({}, railProps)), /*#__PURE__*/(0, _jsxRuntime.jsx)(TrackSlot, (0, _extends2.default)({}, trackProps)), marks.filter(mark => mark.value >= min && mark.value <= max).map((mark, index) => {
      const percent = (0, _useSlider.valueToPercent)(mark.value, min, max);
      const style = axisProps[axis].offset(percent);
      let markActive;
      if (track === false) {
        markActive = values.indexOf(mark.value) !== -1;
      } else {
        markActive = track === 'normal' && (range ? mark.value >= values[0] && mark.value <= values[values.length - 1] : mark.value <= values[0]) || track === 'inverted' && (range ? mark.value <= values[0] || mark.value >= values[values.length - 1] : mark.value >= values[0]);
      }
      return /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(MarkSlot, (0, _extends2.default)({
          "data-index": index
        }, markProps, !(0, _base.isHostComponent)(MarkSlot) && {
          markActive
        }, {
          style: (0, _extends2.default)({}, style, markProps.style),
          className: (0, _clsx.default)(markProps.className, markActive && classes.markActive)
        })), mark.label != null ? /*#__PURE__*/(0, _jsxRuntime.jsx)(MarkLabelSlot, (0, _extends2.default)({
          "aria-hidden": true,
          "data-index": index
        }, markLabelProps, !(0, _base.isHostComponent)(MarkLabelSlot) && {
          markLabelActive: markActive
        }, {
          style: (0, _extends2.default)({}, style, markLabelProps.style),
          className: (0, _clsx.default)(classes.markLabel, markLabelProps.className, markActive && classes.markLabelActive),
          children: mark.label
        })) : null]
      }, index);
    }), values.map((value, index) => {
      const percent = (0, _useSlider.valueToPercent)(value, min, max);
      const style = axisProps[axis].offset(percent);
      const isThumbOverlapping = overlapApi.getIsThumbOverlapping(index, lastActiveThumbIndexRef.current);
      const isValueLabelOverlapping = valueLabelDisplay === 'on' && overlapApi.getIsValueLabelOverlapping(index, lastActiveThumbIndexRef.current);
      return /*#__PURE__*/(0, _jsxRuntime.jsxs)(ThumbSlot, (0, _extends2.default)({
        "data-index": index
      }, thumbProps, {
        className: (0, _clsx.default)(classes.thumb, thumbProps.className, active === index && classes.active, focusedThumbIndex === index && classes.focusVisible, isThumbOverlapping && classes.thumbOverlap),
        style: (0, _extends2.default)({}, style, {
          pointerEvents: disableSwap && active !== index ? 'none' : undefined,
          zIndex: isValueLabelOverlapping || isThumbOverlapping ? 2 : undefined
        }, thumbProps.style),
        ref: thumbRef => {
          var _thumbProps$ref;
          (_thumbProps$ref = thumbProps.ref) == null || _thumbProps$ref.call(thumbProps, thumbRef);
          overlapApi.setThumbRef(index, thumbRef);
        }
      }, !(0, _base.isHostComponent)(ThumbSlot) && {
        isOverlapping: isThumbOverlapping
      }, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(InputSlot, (0, _extends2.default)({
          "data-index": index,
          "aria-label": getAriaLabel ? getAriaLabel(index) : ariaLabel,
          "aria-valuenow": scale(value),
          "aria-labelledby": ariaLabelledby,
          "aria-valuetext": getAriaValueText ? getAriaValueText(scale(value), index) : ariaValuetext,
          value: values[index]
        }, inputSliderProps)), valueLabelDisplay !== 'off' ? /*#__PURE__*/(0, _jsxRuntime.jsx)(ValueLabelSlot, (0, _extends2.default)({}, !(0, _base.isHostComponent)(ValueLabelSlot) && {
          valueLabelFormat,
          valueLabelDisplay,
          value: typeof valueLabelFormat === 'function' ? valueLabelFormat(scale(value), index) : valueLabelFormat,
          disabled,
          index,
          open: open === index || active === index || valueLabelDisplay === 'on',
          isOverlapping: isValueLabelOverlapping
        }, valueLabelProps, {
          className: (0, _clsx.default)(valueLabelProps.className, (open === index || active === index || valueLabelDisplay === 'on') && classes.valueLabelOpen, isValueLabelOverlapping && classes.valueLabelOverlap),
          ref: valueLabelRef => {
            var _valueLabelProps$ref;
            (_valueLabelProps$ref = valueLabelProps.ref) == null || _valueLabelProps$ref.call(valueLabelProps, valueLabelRef);
            overlapApi.setValueLabelRef(index, valueLabelRef);
          },
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
            className: classes.valueLabelLabel,
            children: typeof valueLabelFormat === 'function' ? valueLabelFormat(scale(value), index) : valueLabelFormat
          })
        })) : null]
      }), index);
    })]
  }));
});
process.env.NODE_ENV !== "production" ? Slider.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // |     To update them edit TypeScript types and run "yarn proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * The label of the slider.
   */
  'aria-label': (0, _utils.chainPropTypes)(_propTypes.default.string, props => {
    const range = Array.isArray(props.value || props.defaultValue);
    if (range && props['aria-label'] != null) {
      return new Error('MUI: You need to use the `getAriaLabel` prop instead of `aria-label` when using a range slider.');
    }
    return null;
  }),
  /**
   * The id of the element containing a label for the slider.
   */
  'aria-labelledby': _propTypes.default.string,
  /**
   * A string value that provides a user-friendly name for the current value of the slider.
   */
  'aria-valuetext': (0, _utils.chainPropTypes)(_propTypes.default.string, props => {
    const range = Array.isArray(props.value || props.defaultValue);
    if (range && props['aria-valuetext'] != null) {
      return new Error('MUI: You need to use the `getAriaValueText` prop instead of `aria-valuetext` when using a range slider.');
    }
    return null;
  }),
  /**
   * Override or extend the styles applied to the component.
   */
  classes: _propTypes.default.object,
  /**
   * @ignore
   */
  className: _propTypes.default.string,
  /**
   * The color of the component.
   * It supports both default and custom theme colors, which can be added as shown in the
   * [palette customization guide](https://mui.com/material-ui/customization/palette/#adding-new-colors).
   * @default 'primary'
   */
  color: _propTypes.default /* @typescript-to-proptypes-ignore */.oneOfType([_propTypes.default.oneOf(['primary', 'secondary']), _propTypes.default.string]),
  /**
   * The default value. Use when the component is not controlled.
   */
  defaultValue: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.number]),
  /**
   * If `true`, the component is disabled.
   * @default false
   */
  disabled: _propTypes.default.bool,
  /**
   * If `true`, the active thumb doesn't swap when moving pointer over a thumb while dragging another thumb.
   * @default false
   */
  disableSwap: _propTypes.default.bool,
  /**
   * Accepts a function which returns a string value that provides a user-friendly name for the thumb labels of the slider.
   * This is important for screen reader users.
   * @param {number} index The thumb label's index to format.
   * @returns {string}
   */
  getAriaLabel: _propTypes.default.func,
  /**
   * Accepts a function which returns a string value that provides a user-friendly name for the current value of the slider.
   * This is important for screen reader users.
   * @param {number} value The thumb label's value to format.
   * @param {number} index The thumb label's index to format.
   * @returns {string}
   */
  getAriaValueText: _propTypes.default.func,
  /**
   * Marks indicate predetermined values to which the user can move the slider.
   * If `true` the marks are spaced according the value of the `step` prop.
   * If an array, it should contain objects with `value` and an optional `label` keys.
   * @default false
   */
  marks: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.shape({
    label: _propTypes.default.node,
    value: _propTypes.default.number.isRequired
  })), _propTypes.default.bool]),
  /**
   * The maximum allowed value of the slider.
   * Should not be equal to min.
   * @default 100
   */
  max: _propTypes.default.number,
  /**
   * The minimum allowed value of the slider.
   * Should not be equal to max.
   * @default 0
   */
  min: _propTypes.default.number,
  /**
   * Name attribute of the hidden `input` element.
   */
  name: _propTypes.default.string,
  /**
   * Callback function that is fired when the slider's value changed.
   *
   * @param {Event} event The event source of the callback.
   * You can pull out the new value by accessing `event.target.value` (any).
   * **Warning**: This is a generic event, not a change event.
   * @param {number | number[]} value The new value.
   * @param {number} activeThumb Index of the currently moved thumb.
   */
  onChange: _propTypes.default.func,
  /**
   * Callback function that is fired when the `mouseup` is triggered.
   *
   * @param {React.SyntheticEvent | Event} event The event source of the callback. **Warning**: This is a generic event, not a change event.
   * @param {number | number[]} value The new value.
   */
  onChangeCommitted: _propTypes.default.func,
  /**
   * The component orientation.
   * @default 'horizontal'
   */
  orientation: _propTypes.default.oneOf(['horizontal', 'vertical']),
  /**
   * A transformation function, to change the scale of the slider.
   * @param {any} x
   * @returns {any}
   * @default function Identity(x) {
   *   return x;
   * }
   */
  scale: _propTypes.default.func,
  /**
   * The size of the slider.
   * @default 'medium'
   */
  size: _propTypes.default /* @typescript-to-proptypes-ignore */.oneOfType([_propTypes.default.oneOf(['small', 'medium']), _propTypes.default.string]),
  /**
   * The props used for each slot inside the Slider.
   * @default {}
   */
  slotProps: _propTypes.default.shape({
    input: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
    mark: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
    markLabel: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
    rail: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
    root: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
    thumb: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
    track: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
    valueLabel: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object])
  }),
  /**
   * The components used for each slot inside the Slider.
   * Either a string to use a HTML element or a component.
   * @default {}
   */
  slots: _propTypes.default.shape({
    input: _propTypes.default.elementType,
    mark: _propTypes.default.elementType,
    markLabel: _propTypes.default.elementType,
    rail: _propTypes.default.elementType,
    root: _propTypes.default.elementType,
    thumb: _propTypes.default.elementType,
    track: _propTypes.default.elementType,
    valueLabel: _propTypes.default.elementType
  }),
  /**
   * The granularity with which the slider can step through values. (A "discrete" slider.)
   * The `min` prop serves as the origin for the valid values.
   * We recommend (max - min) to be evenly divisible by the step.
   *
   * When step is `null`, the thumb can only be slid onto marks provided with the `marks` prop.
   * @default 1
   */
  step: _propTypes.default.number,
  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object, _propTypes.default.bool])), _propTypes.default.func, _propTypes.default.object]),
  /**
   * Tab index attribute of the hidden `input` element.
   */
  tabIndex: _propTypes.default.number,
  /**
   * The track presentation:
   *
   * - `normal` the track will render a bar representing the slider value.
   * - `inverted` the track will render a bar representing the remaining slider value.
   * - `false` the track will render without a bar.
   * @default 'normal'
   */
  track: _propTypes.default.oneOf(['inverted', 'normal', false]),
  /**
   * The value of the slider.
   * For ranged sliders, provide an array with two values.
   */
  value: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.number), _propTypes.default.number]),
  /**
   * Controls when the value label is displayed:
   *
   * - `auto` the value label will display when the thumb is hovered or focused.
   * - `on` will display persistently.
   * - `off` will never display.
   * @default 'off'
   */
  valueLabelDisplay: _propTypes.default.oneOf(['auto', 'off', 'on']),
  /**
   * The format function the value label's value.
   *
   * When a function is provided, it should have the following signature:
   *
   * - {number} value The value label's value to format
   * - {number} index The value label's index to format
   * @param {any} x
   * @returns {any}
   * @default function Identity(x) {
   *   return x;
   * }
   */
  valueLabelFormat: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string])
} : void 0;
var _default = exports.default = Slider;
