// @mui/material-next/node/Slider/useSliderElementsOverlap.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useSliderElementsOverlap;
var React = _interopRequireWildcard(require("react"));
var _utils = require("@mui/utils");
var _lodash = require("lodash");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const overlapCompareFunctionByAxis = {
  horizontal: (firstElementRect, secondElementRect, margin) => firstElementRect.right + margin > secondElementRect.left,
  'horizontal-reverse': (firstElementRect, secondElementRect, margin) => secondElementRect.right + margin > firstElementRect.left,
  vertical: (firstElementRect, secondElementRect, margin) => secondElementRect.bottom + margin > firstElementRect.top
};
function getSliderElementsOverlap(elements, axis, margin) {
  const overlapArray = elements.map(() => false);
  for (let elementIndex = 0; elementIndex < elements.length - 1; elementIndex += 1) {
    const firstElementRect = elements[elementIndex].getBoundingClientRect();
    const secondElementRect = elements[elementIndex + 1].getBoundingClientRect();
    if (overlapCompareFunctionByAxis[axis](firstElementRect, secondElementRect, margin)) {
      overlapArray[elementIndex] = true;
      overlapArray[elementIndex + 1] = true;
    }
  }
  return overlapArray;
}
function useElementsOverlap(axis, margin = 0) {
  const elementsRefList = React.useRef([]);
  const [elementsOverlapArray, setElementsOverlapArray] = React.useState([]);
  const setRef = React.useCallback((elementIndex, ref) => {
    if (!!ref && elementsRefList.current[elementIndex] !== ref) {
      elementsRefList.current[elementIndex] = ref;
    }
  }, []);
  const getIsOverlapping = React.useCallback((elementIndex, lastActiveIndex) => {
    if (elementsRefList.current.length < 2) {
      return false;
    }
    return elementsOverlapArray[elementIndex] && lastActiveIndex === elementIndex;
  }, [elementsOverlapArray]);
  const onMove = React.useCallback(() => {
    if (elementsRefList.current.length > 1) {
      const updatedOverlap = getSliderElementsOverlap(elementsRefList.current, axis, margin);
      if (!(0, _lodash.isEqual)(updatedOverlap, elementsOverlapArray)) {
        setElementsOverlapArray(updatedOverlap);
      }
    }
  }, [axis, elementsOverlapArray, margin]);
  return {
    setRef,
    getIsOverlapping,
    onMove
  };
}
function useSliderElementsOverlap(axis) {
  const {
    setRef: setThumbRef,
    getIsOverlapping: getIsThumbOverlapping,
    onMove: onThumbMove
  } = useElementsOverlap(axis);

  // ValueLabel -12px margin is required due to how its "inverted water drop"
  // shape is built with CSS. Might want to allow this to be configurable in the future.
  const {
    setRef: setValueLabelRef,
    getIsOverlapping: getIsValueLabelOverlapping,
    onMove: onValueLabelMove
  } = useElementsOverlap(axis, -12);
  const onThumbMoved = React.useCallback(() => {
    onThumbMove();
    onValueLabelMove();
  }, [onThumbMove, onValueLabelMove]);
  const debouncedOnThumbMoved = React.useMemo(() => (0, _lodash.debounce)(onThumbMoved, 50), [onThumbMoved]);
  (0, _utils.unstable_useEnhancedEffect)(() => {
    onThumbMoved();
  }, [onThumbMoved]);
  return {
    setThumbRef,
    setValueLabelRef,
    getIsThumbOverlapping,
    getIsValueLabelOverlapping,
    onThumbMoved: debouncedOnThumbMoved
  };
}
